src/modules/python/pbs_hooks/PBS_power.PY

```diff
@@ -76,16 +80,308 @@ def vnodes_enabled(job):
+    except:
+        return None
+    aliaslist.append(hostname)
+    # Search for possible match in server vnode list.
+    pbsvnodes = dict()
+    for vn in pbs.server().vnodes():
+        pbsvnodes[vn.name] = vn
+    for n in aliaslist:
+         if n in pbsvnodes:
+             return pbsvnodes[n]
+    return None
+
+
+# Read the config file in json format
+def parse_config_file():
+    # Turn everything off by default. These settings be modified
+    # when the configuration file is read.
+    global pbs_home
+    global pbs_exec
+    global power_ramp_rate_enable
+    global power_on_off_enable
+    global node_idle_limit
+    global min_node_down_delay
+    global max_jobs_analyze_limit
+    global max_concurrent_nodes
+
+    try:
+    # This block will work for PBS Pro versions 13 and later
+        pbs_conf = pbs.get_pbs_conf()
+        pbs_home = pbs_conf['PBS_HOME']
+        pbs_exec = pbs_conf['PBS_EXEC']
+    except:
+        pbs.logmsg(pbs.EVENT_DEBUG, "PBS_HOME needs to be defined in the config file")
+        pbs.logmsg(pbs.EVENT_DEBUG, "Exiting the power hook")
+        pbs.event().accept()
+
+    # Identify the config file and read in the data
+    config_file = ''
+    if 'PBS_HOOK_CONFIG_FILE' in os.environ:
+        config_file = os.environ["PBS_HOOK_CONFIG_FILE"]
+    tmpcfg = ''
+    if not config_file:
+        tmpcfg = os.path.join(pbs_home, 'server_priv', 'hooks',
+			  'PBS_power.CF')
+    if os.path.isfile(tmpcfg):
+        config_file = tmpcfg
+    if not config_file:
+        tmpcfg = os.path.join(pbs_home, 'mom_priv', 'hooks',
+			  'PBS_power.CF')
+    if os.path.isfile(tmpcfg):
+        config_file = tmpcfg
+    if not config_file:
+        raise Exception("Config file not found")
+    pbs.logmsg(pbs.EVENT_DEBUG4, "Config file is %s" % config_file)
+    try:
+        fd = open(config_file, 'r')
+        config = json.load(fd)
+        fd.close()
+    except IOError:
+        raise Exception("I/O error reading config file")
+    except:
+        raise Exception("Error reading config file")
+
+    #Assign default values to attributes
+    power_ramp_rate_enable = False
+    power_on_off_enable = False
+    node_idle_limit = 1800
+    min_node_down_delay = 1800
+    max_jobs_analyze_limit = 100
+    max_concurrent_nodes = 10
+
+    #Now assgin values read from config file
+    if 'power_on_off_enable' in config:
+        power_on_off_enable = config['power_on_off_enable']
+    if 'power_ramp_rate_enable' in config:
+        power_ramp_rate_enable = config['power_ramp_rate_enable']
+    if 'node_idle_limit' in config:
+        node_idle_limit = int(config['node_idle_limit'])
+    if 'min_node_down_delay' in config:
+        min_node_down_delay = int(config['min_node_down_delay'])
+    if 'max_jobs_analyze_limit' in config:
+        max_jobs_analyze_limit = int(config['max_jobs_analyze_limit'])
+    if 'max_concurrent_nodes' in config:
+        max_concurrent_nodes = int(config['max_concurrent_nodes'])
+
```

Comment:
Can we log a msg into logs , When some attribute gets updated ?

Reply:
Can we validate values and assign to default if the value is wrong .

Reply:
Fixed.